kernel/univ.ml

```diff
@@ -942,17 +942,30 @@ let hcons_universe_context = UContext.hcons
 
 module AUContext =
 struct
-  include UContext
+  type t = Names.Name.t list constrained
 
   let repr (inst, cst) =
-    (Array.mapi (fun i _l -> Level.var i) inst, cst)
+    (Array.init (List.length inst) (fun i -> Level.var i), cst)
 
-  let pr f ?variance ctx = pr f ?variance (repr ctx)
+  let pr f ?variance ctx = UContext.pr f ?variance (repr ctx)
 
   let instantiate inst (u, cst) =
-    assert (Array.length u = Array.length inst);
+    assert (List.length u = Array.length inst);
     subst_instance_constraints inst cst
 
+  let names (nas, _) = nas
+
+  let hcons (univs, cst) =
+    (List.map Names.Name.hcons univs, hcons_constraints cst)
```

Comment:
Shouldn't this do something on the list as a whole too? Although the way we use hcons is so adhoc it's hard to know anything.

Reply:
Indeed, but that would require writing yet another table. Note that this is not the only case where we don't properly hashcons intermediate structures for the sake of simplicity.

Reply:
There's some functor magic used by Universe, so it would just be ~2 lines.
Maybe we shouldn't hcons at all this part.